### PR TITLE
Ensure that Parameters compare equal when they have the same hash

### LINKED ISSUE #11654: Parameter instances  can compare equal while having different hash values
### Environment

- **Qiskit Terra version**: 0.45.2
- **Python version**: 3.12
- **Operating system**: Fedora Linux 39

### What is happening?

Currently, two `Parameter` instances can compare equal while not having the same hash value. This condition violates the Python data model which is [documented](https://docs.python.org/3/reference/datamodel.html#object.__hash__) as requiring:

> The only required property is that objects which compare equal have the same hash value

### How can we reproduce the issue?

```python
a = Parameter("a")
b = Parameter("b", uuid=a._uuid)

a == b  # True
hash(a) == hash(b)  # False
```

### What should happen?

Here there is a decision to make. Parameters are essentially a tuple of a name string and a Python UUID. Currently, `__eq__` uses only the UUID while `__hash__` uses the name and the UUID. The two methods should use the same data.

In https://github.com/Qiskit/qiskit/pull/11647#issuecomment-1912087645, @jakelishman mentioned that the documentation states that the name is required for equality, so I created #11652 to add the name to the equality check (I think @jakelishman was referring to "Setting the ``uuid`` field when creating two parameters to the same thing (along with the same name) allows them to be equal." in the `Parameter.__init__` docstring). However, in the pulse module, there is code that makes use of the fact that parameter names are not part of the equality test by redefining the name:

https://github.com/Qiskit/qiskit/blob/ce02f068ae1cf919373709e74a3dfbb6be3a2aa2/qiskit/pulse/schedule.py#L882-L912

So we have to decide which way to go:

1. Remove `name` from the hash calculation so that only the UUID matters for Parameter instances. In practice, this is probably fine since the UUID's should always be unique.
2. Add `name` to the equality test and rework the pulse code. It is just using the renaming for convenience of labeling parameters. Perhaps the labeling could be done differently (`scoped_parameters` could return a dict with the keys being the scoped parameter names?). I don't see `scoped_parameters` used anywhere here or qiskit-experiments currently.

I lean toward option 2. I think a key question is how tied is `Parameter` to `ParameterExpression` (technically it is a direct subclass so very tied :slightly_smiling_face:). If `Parameter` just needs to be a variable that floats around through code objects and can later be assigned a value, allowing different names seems fine. However, if it is important that a `Parameter` can always be used in an expression and then later substituted for a different value, allowing different names is a problem because `ParameterExpression` maps the names directly to sympy/symengine symbols and wants to substitute them using the sympy/symengine mechanism. If `Parameter`s are being passed around and manipulated, you could end up calling `assign` with a `Parameter` instance with a different name from the one that was used to create the underlying symbolic expression of a `ParameterExpression`. Or you could say that both use cases are allowed -- when doing simple value substitution renaming `Parameter`s is okay; when doing symbolic expressions, you have to be careful not to reuse UUID's with different names.

@nkanazawa1989 What do you think as the author of the scoped_parameters code?

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by wshanks**
Previously, only the `Parameter.uuid` was compared, so `Paramemter`
instances with different names could compare equal if they had been
constructed using a common value for the `uuid` parameter (which is
usually not passed explicitly).

Fix #11654

### CODE REVIEW

**Comment by wshanks** [comment]
Currently, I based this on #11647 so that I could add to the test from there and reference `uuid` in Sphinx. This can be reworked if #11647 can only go in to 1.1 and this PR is wanted in 1.0 as a bug fix.

**Comment by wshanks** [comment]
It looks like there are some existing uses of different names comparing equal so this one may take some more careful work.

**Comment by wshanks** [comment]
This will wait for a resolution in #11654

File: qiskit/primitives/containers/bindings_array.py
Comment (jakelishman) [inline qiskit/primitives/containers/bindings_array.py:1]: Bad merge?

File: releasenotes/notes/parameter-uuid-60dd46a739afabce.yaml
Comment (jakelishman) [inline releasenotes/notes/parameter-uuid-60dd46a739afabce.yaml:1]: It doesn't matter at all, but for the future: you can put more than one bullet (and even more than one section) in a release note.

File: test/python/primitives/containers/test_bindings_array.py
Comment (jakelishman) [inline test/python/primitives/containers/test_bindings_array.py:1]: Bad merge?

**Review by jakelishman — COMMENTED** [review]
Thanks Will - pending resolution of #11654 this looks good to go, except for maybe there was a slightly dodgy merge somewhere?

File: qiskit/primitives/containers/bindings_array.py
Comment (wshanks) [reply qiskit/primitives/containers/bindings_array.py:1]: Not sure what happened here, but I will clean it up. I must have done something funny splitting this change out from the `uuid` property which I had together originally.

Comment (wshanks) [reply qiskit/primitives/containers/bindings_array.py:1]: Oh, I wanted to use this with the changes in #11647 before that one was merged, so this was based on it. That is why it had the two release notes files for example. I am not sure how I got the extra binding array commit in in addition to the #11647 commits.

**Comment by wshanks** [comment]
https://github.com/Qiskit/qiskit/pull/11692 modifies the tests that are currently failing here so that they are not affected by this change.

**Comment by jakelishman** [comment]
In the interests of time, and because this PR has an extended critical path to merge, I'm going to defer it to 1.0.1.  It's a bugfix - the current definition of equality is broken and not what's documented or compatible with the hash.

**Review by jakelishman — APPROVED** [review]
Thanks Will, this looks good to me, and it's great to fix a data-model problem like this.

**Review by ihincks — APPROVED** [review]
Thanks @wshanks

**Pull request merged**